### PR TITLE
Improve dav

### PR DESCRIPTION
**Pull request by MrGumpIT**
- handles the exception due to ` [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: PathAccessException: Deletion failed, path = 'C:\Users\user\OneDrive\Documenti\Linwood\Butterfly\Remotes\Butterfly\Templates' (OS Error:  The file could not be accessed. The file is being used by another process.
, errno = 32)
0      _checkForErrorResponse (dart:io/common.dart:55:9)
1      _Directory._delete.<anonymous closure> (dart:io/directory_impl.dart:180:7) <asynchronous suspension>
2      RemoteSystem.clearCachedContent (package:butterfly/api/file_system/file_system_remote.dart:149:7)
<asynchronous suspension>
`

- When you created a file without specifying its name or using Quick Launch, you would get a conflict, which led to the creation of multiple folders instead of the file

### CODE REVIEW

File: app/lib/api/file_system/file_system_dav.dart [hunk 1]
```diff
@@ -202,20 +202,46 @@ class DavRemoteDocumentFileSystem extends DocumentRemoteSystem {
...
     final directoryPath = path.substring(0, last);
-    if (!await hasAsset(directoryPath)) {
-      await createDirectory(directoryPath);
-    }
-    final response =
-        await createRequest(path.split('/'), method: 'PUT', bodyBytes: data);
-    if (response?.statusCode != 201 && response?.statusCode != 204) {
-      throw Exception(
-          'Failed to update document: ${response?.statusCode} ${response?.reasonPhrase}');
+    final directory = Directory(directoryPath);
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_dav.dart:213]: I think we shouldn't use dart:io for remote storage. Maybe this directory isn't synced?

File: app/lib/api/file_system/file_system_dav.dart [hunk 2]
```diff
@@ -202,20 +202,46 @@ class DavRemoteDocumentFileSystem extends DocumentRemoteSystem {
...
+    if (!await directory.exists()) {
+      await directory.create(recursive: true);
+    }
+
+    String newPath = path;
+    int counter = 1;
+    bool fileUpdated = false;
+
+    while (!fileUpdated) {
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_dav.dart:222]: Why a while? We should only try it once

File: app/lib/api/file_system/file_system_remote.dart [hunk 3]
```diff
@@ -144,9 +144,29 @@ mixin RemoteSystem {
 
   Future<void> clearCachedContent() async {
     var cacheDir = await getRemoteCacheDirectory();
+    print('Cache directory path: $cacheDir');
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_remote.dart:147]: we don't use prints inside the app

File: app/lib/api/file_system/file_system_remote.dart [hunk 4]
```diff
@@ -144,9 +144,29 @@ mixin RemoteSystem {
...
+    if (exists) {
+      try {
+        await directory.delete(recursive: true);
+        print('Directory deleted successfully.');
+      } on FileSystemException catch (e) {
+        if (e.osError?.errorCode == 32) {
+          print('Directory is in use, retrying... $e');
+          // Retry after a short delay
+          await Future.delayed(const Duration(seconds: 5));
+          await clearCachedContent();
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_remote.dart:160]: maybe a while with a limit is better?
Otherwise it can happen that you will get a stack overflow exception

File: app/lib/api/file_system/file_system_remote.dart [hunk 5]
```diff
@@ -144,9 +144,29 @@ mixin RemoteSystem {
 
   Future<void> clearCachedContent() async {
     var cacheDir = await getRemoteCacheDirectory();
+    print('Cache directory path: $cacheDir');
     var directory = Directory(cacheDir);
-    if (await directory.exists()) {
-      await directory.delete(recursive: true);
+    bool exists = await directory.exists();
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_remote.dart:149]:
```suggestion
    final exists = await directory.exists();
```
like said in the pdf import pr, a final is better since it indicates that this value won't be changed

**Review by CodeDoctorDE — CHANGES_REQUESTED** [review]
thanks for this pr, I really like your changes, I just have some adjustments i would make

File: app/lib/api/file_system/file_system_remote.dart [hunk 6]
```diff
@@ -143,10 +143,30 @@ mixin RemoteSystem {
...
+      final exists = await directory.exists();
+      if (exists) {
+        try {
+          await directory.delete(recursive: true);
+          // Directory deleted successfully, perform any needed actions
+        } on FileSystemException catch (e) {
+          if (e.osError?.errorCode == 32) {
+            // Directory in use, retry after a short delay
+            await Future.delayed(const Duration(seconds: 1));
+            await clearCachedContent();
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_remote.dart:158]: Here I wouldn't add recursive since it can result in a stack overflow. Maybe better a for/while loop with a limit of 5?

File: app/lib/api/file_system/file_system_remote.dart [hunk 7]
```diff
@@ -143,10 +143,30 @@ mixin RemoteSystem {
...
+          } else if (e.osError?.errorCode == 2) {
+            // Directory not found, no further action needed
+          } else {
+            // Handle unexpected FileSystemException
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_remote.dart:162]: maybe rethrow it to show it to the user?

File: app/lib/api/file_system/file_system_remote.dart [hunk 8]
```diff
@@ -143,10 +143,30 @@ mixin RemoteSystem {
...
+          }
+        }
+      } else {
+        // No directory exists, no action needed
+      }
+    } catch (e) {
+      // Handle any other exceptions that may occur
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_remote.dart:169]: I wouldn't put it in a try catch, the method that called this function should know that there is an exception happened

File: app/lib/api/file_system/file_system_dav.dart [hunk 9]
```diff
@@ -202,18 +202,36 @@ class DavRemoteDocumentFileSystem extends DocumentRemoteSystem {
...
-    if (response?.statusCode != 201 && response?.statusCode != 204) {
+        await createRequest(p.split(path), method: 'PUT', bodyBytes: data);
+    if (response?.statusCode == 201 || response?.statusCode == 204) {
+      // File updated
+      return;
+    } else if (response?.statusCode == 409) {
+      // Conflict, rename the file
+      final baseName = p.basenameWithoutExtension(path);
+      final extension = p.extension(path);
+      String newPath = p.join(directoryPath, '$baseName(1)$extension');
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_dav.dart:222]: This is a "updateFile" function. It should override the path if it already exists

**Review by CodeDoctorDE — CHANGES_REQUESTED** [review]
thanks for your hard work, I reviewed your code, feel free to comment on these if you disagree

File: app/lib/api/file_system/file_system_dav.dart (hunk 9 above)
Comment (MrGumpIT) [reply app/lib/api/file_system/file_system_dav.dart:222]: no you need to rename, because when you create a file using quick launch and don't give it a name it leads to errors

File: app/lib/api/file_system/file_system_remote.dart (hunk 6 above)
Comment (MrGumpIT) [reply app/lib/api/file_system/file_system_remote.dart:158]: I didn't understand before, you're right. I'll edit it

File: app/lib/api/file_system/file_system_dav.dart (hunk 9 above)
Comment (CodeDoctorDE) [reply app/lib/api/file_system/file_system_dav.dart:222]: That is been handled inside the create file method: https://github.com/LinwoodDev/Butterfly/blob/a79aa7777f6a3a9a0768567cf7292520d3c40bb9/app/lib/api/file_system/file_system.dart#L180

It tries to find the next unique name (`My file.bfly`, `My file (1).bfly`, `My file (2).bfly`) and runs `updateFile` after that.

But `updateFile` will also be run if you open an existing document and click on the save button. Here we want to override the current file.

So in conclusion we want `updateFile` to always override the current file if it exists.

Comment (MrGumpIT) [reply app/lib/api/file_system/file_system_dav.dart:222]: ok then I avoid managing 409 to rename the file. it actually works

Comment (MrGumpIT) [reply app/lib/api/file_system/file_system_dav.dart:222]: i did it like this ` @override
  Future<void> updateFile(String path, List<int> data,
      {bool forceSync = false}) async {
    // Creazione della directory se non esiste
    final directoryPath = p.dirname(path);
    final directory = Directory(directoryPath);
    if (!await directory.exists()) {
      await directory.create(recursive: true);
    }

    // Request to overwrite the file
    final response =
        await createRequest(p.split(path), method: 'PUT', bodyBytes: data);
    if (response?.statusCode == 200 ||
        response?.statusCode == 201 ||
        response?.statusCode == 204) {
      // File overwritten successfully
      return;
    } else {
      // Management of error cases
      throw Exception(
          'Failed to update document: ${response?.statusCode} ${response?.reasonPhrase}');
    }
  }` works correctly, except in the case where the user deliberately creates a file with the same name as another file, in which case you get `Unhandled Exception: Exception: Failed to update document: 500 Internal Server Error` perhaps the best way would be to not handle it and warn the user with a popup that they need to rename the file

Comment (CodeDoctorDE) [reply app/lib/api/file_system/file_system_dav.dart:222]: Yeah maybe, but which webdav server do you use? On nextcloud it works like a charm

Comment (MrGumpIT) [reply app/lib/api/file_system/file_system_dav.dart:222]: i use nextcloud

File: app/lib/api/file_system/file_system_dav.dart (hunk 1 above)
Comment (CodeDoctorDE) [reply app/lib/api/file_system/file_system_dav.dart:204]: Why did you resolve this conversation?
I would readd the cache system:
```dart
if(remote.hasDocumentCached(path)) {
      cacheContent(path, data);
    }
```
but why the directory code here? The path inside the parameter has nothing to do with the cached on the disk.

Comment (MrGumpIT) [reply app/lib/api/file_system/file_system_dav.dart:204]: the directory code was already there

Comment (CodeDoctorDE) [reply app/lib/api/file_system/file_system_dav.dart:204]: No the code was:
```dart
final directoryPath = path.substring(0, last);
    if (!await hasAsset(directoryPath)) {
      await createDirectory(directoryPath);
```

This checks remotly if the directory exists and create them remotly

File: app/lib/api/file_system/file_system_dav.dart [hunk 10]
```diff
@@ -199,11 +199,20 @@ class DavRemoteDocumentFileSystem extends DocumentRemoteSystem {
...
-    if (!await directory.exists()) {
-      await directory.create(recursive: true);
+    // Cache check
+    if (!forceSync && remote.hasDocumentCached(path)) {
+      cacheContent(path, data);
+      return;
+    }
+
+    // Create directory if not exits
+    var last = path.lastIndexOf('/');
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_dav.dart:209]: Hmm i liked the p.dirname since it does handle things like `/directory/subdirectory/`.

File: app/lib/api/file_system/file_system_dav.dart [hunk 11]
```diff
@@ -199,11 +199,20 @@ class DavRemoteDocumentFileSystem extends DocumentRemoteSystem {
...
       {bool forceSync = false}) async {
-    // Create directory if not exists
-    final directoryPath = p.dirname(path);
-    final directory = Directory(directoryPath);
-    if (!await directory.exists()) {
-      await directory.create(recursive: true);
+    // Cache check
+    if (!forceSync && remote.hasDocumentCached(path)) {
+      cacheContent(path, data);
+      return;
```
Comment (CodeDoctorDE) [inline app/lib/api/file_system/file_system_dav.dart:205]: This was a mistake on my side, we shouldn't return when it should be cached, it should also update it on the server

**Review by CodeDoctorDE — CHANGES_REQUESTED** [review]

File: app/lib/api/file_system/file_system_dav.dart (hunk 10 above)
Comment (CodeDoctorDE) [reply app/lib/api/file_system/file_system_dav.dart:209]: I didn't know about this dirname function before. Maybe we should add more of these useful path package functions in butterfly to improve error handling like trailing slashes?

File: app/lib/api/file_system/file_system_dav.dart (hunk 11 above)
Comment (MrGumpIT) [reply app/lib/api/file_system/file_system_dav.dart:205]: okay, so you want to go back to this? 
`Future<void> updateFile(String path, List<int> data,
    {bool forceSync = false}) async {
  if (!forceSync && remote.hasDocumentCached(path)) {
    cacheContent(path, data);
  }`

File: app/lib/api/file_system/file_system_dav.dart (hunk 10 above)
Comment (MrGumpIT) [reply app/lib/api/file_system/file_system_dav.dart:209]: I can try, but I would like to solve the rest first. because I'm getting lost with all the comments 😂

File: app/lib/api/file_system/file_system_dav.dart (hunk 11 above)
Comment (CodeDoctorDE) [reply app/lib/api/file_system/file_system_dav.dart:205]: Just remove the return line and it should work just fine

File: app/lib/api/file_system/file_system_dav.dart (hunk 10 above)
Comment (CodeDoctorDE) [reply app/lib/api/file_system/file_system_dav.dart:209]: Sure but this and the return line removal are just the changes I would suggest, otherwise it loosk good!

**Review by CodeDoctorDE — APPROVED** [review]
lgtm, will merge it tomorrow!

**Comment by MrGumpIT** [comment]
thanks, I will try to implement p. subsequently
